### PR TITLE
Model ladder, revamped

### PR DESCRIPTION
**Pull request by epwalsh**
Adds a new model ladder API following [this design doc](https://docs.google.com/document/d/1pCdq1b6kkw81n3WNY1rq1CHgtotG2hIVdeZPLPLnXYA/edit?tab=t.0) and a script to run the baseline Olmo 3 ladder, which also serves an example to configure new ladder experiments.

~~For now the new ladder API is located at `src/olmo_core/model_ladder2/` to make the diffs easier to review, but before merging we'll replace the old code at `src/olmo_core/model_ladder.py` with this new module.~~

You can play around with the baseline ladder script without actually launching anything via its `dry-run` command. For example:

```bash
python src/scripts/train/ladder/olmo3_ladder.py dry-run --size=760M --show-plot
```

---

> [!NOTE]
> Replaces the old ladder with a new modular API (configurators + CLI), adds an Olmo3 ladder script and CI dry-runs, and updates deps.
> 
> - **Model Ladder (revamp)**:
>   - New modular API under `olmo_core/model_ladder/` with `ModelLadder`, `ModelConfigurator`, `RunConfigurator`, `DeviceMeshSpec`, checkpoint/metrics utilities, and helpers.
>   - Transformer-specific configurator (`TransformerModelConfigurator`, `TransformerSize`) and WSDS+Chinchilla run configurator (`WSDSChinchillaRunConfigurator`), including LR plotting and checkpoint interval logic.
>   - Removes legacy ladder modules `olmo_core/model_ladder.py`, `olmo_core/internal/model_ladder.py`, old OLMo2 ladder script, and its test.
> - **Internal CLI**:
>   - Adds `olmo_core/internal/ladder.py` with subcommands: `dry-run`, `benchmark`, `launch-benchmark`, `run`, `launch`, `launch-all`, `status`, `metrics` (Beaker-integrated).
> - **Script**:
>   - New `scripts/train/ladder/olmo3_ladder.py` configuring a baseline Olmo3 ladder (composable data pipeline, WSDS/Chinchilla scheduling).
> - **CI**:
>   - Workflow runs `dry-run` across sizes (`190M`–`13B`).
> - **Deps**:
>   - Adds `pandas` (core) and `matplotlib` (dev).
> - **Misc**:
>   - `CHANGELOG.md` notes API changes.
>   - Adds `utils.warn_once()` export.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f64b5ac961a3f6b27fecc6f26abf6565686a855c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+from .base import RunConfigurator
+from .utils import format_count, format_tokens
+
+
+@dataclass(kw_only=True)
+class WSDSChinchillaRunConfigurator(RunConfigurator):
+    """
+    A run configurator that uses WSD-S learning rate scheduling and Chinchilla scaling laws.
+    """
+
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py:28]: nit: the underlying chinchilla heuristic of 20 tok/param has some caveats, like it only applies to models trained with AdamW and is dataset dependent (was determined on the Pile iirc). Might be worth adding a disclaimer here so people dont assume that 20 tok/param is actually optimal for their dataset.

Comment (epwalsh) [reply src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py:28]: Done, and made it configurable: 5aa3d4e1

File: src/olmo_core/model_ladder/base.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+        optim_config: OptimConfig,
+        scheduler: Scheduler,
+        device_type: str,
+    ) -> TrainModule:
+        """Build the train module for the given model and optimizer configs."""
+        raise NotImplementedError
+
+
+@dataclass(kw_only=True)
+class RunConfigurator(Config, metaclass=ABCMeta):
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/base.py:150]: Seems like this loosely maps to TrainModule and the above maps to Model?

File: src/olmo_core/model_ladder/base.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+        save_path: PathOrStr | None = None,
+    ) -> PathOrStr | None:
+        """Render a plot of the learning rate schedule."""
+        raise NotImplementedError
+
+
+@dataclass(kw_only=True)
+class ModelLadder(Config):
+    """
+    Represents a complete model ladder of runs.
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/base.py:205]: And this roughly maps to ExperimentConfig

File: src/olmo_core/model_ladder/base.py (hunk 2 above)
Comment (epwalsh) [reply src/olmo_core/model_ladder/base.py:150]: I don't think that's the right comparison because the `ModelConfigurator` is actually responsible for building the `TrainModule` given the optimizer, scheduler, and other hyperparameters configured by the `RunConfigurator`. I'm not entirely happy with it, but I landed there because the training optimizations/parallelism configuration that go into the `TrainModule` are tightly coupled to the model architecture, which is defined by the `ModelConfigurator`.

Comment (epwalsh) [reply src/olmo_core/model_ladder/base.py:150]: I think of `RunConfigurator` as what decides hyperparameters.

File: src/olmo_core/model_ladder2/base.py [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+        size_spec: str,
+        sequence_length: int,
+        tokenizer: TokenizerConfig,
+        device_type: str,
+    ) -> M:
+        """Configure the model for the given size spec."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def configure_device_microbatch_size(
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder2/base.py]: elsewhere we use the term "rank_microbatch_size", can we keep that consistent?

Comment (epwalsh) [reply src/olmo_core/model_ladder2/base.py]: Done: bf5bc2de

File: src/olmo_core/model_ladder/base.py [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        ]
+
+    def _duration_to_steps(self, d: Duration, global_batch_size: int) -> int:
+        if d.unit == DurationUnit.steps:
+            return d.value
+        elif d.unit == DurationUnit.tokens:
+            steps = d.value // global_batch_size
+            return steps
+        else:
+            raise ValueError(f"Unsupported checkpoint interval duration unit: {d.unit}.")
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/base.py:509]: This might make a good method on the duration class

File: src/olmo_core/model_ladder2/base.py [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+                "metric_saver": callbacks.MetricSaverCallback(
+                    fixed_steps=checkpoint_interval_steps,
+                    enabled=not for_benchmarking,
+                ),
+            },
+        )
+
+    def _get_in_loop_eval_tasks(self) -> list[str]:
+        # For training runs where we don't expect the model to acquire MC (e.g., 1B-5xC, short 7B training runs).
+        tasks_small_compute = [
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder2/base.py]: Looks like these are the tasks groups from `task_groups.py` but without the basic_skills and mt_mbpp tasks?

Could we just use the predefined tasks groups instead?

File: src/olmo_core/model_ladder/transformer_model_configurator.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    TransformerTrainModuleConfig,
+)
+from olmo_core.utils import warn_once
+
+from .base import DeviceMeshSpec, ModelConfigurator
+from .utils import format_count
+
+
+class TransformerSize(StrEnum):
+    size_190M = "190M"
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/transformer_model_configurator.py:25]: Just noting to myself that it looks like the smallest size takes about 30m to hit 1xC

File: src/olmo_core/model_ladder2/transformer_model_configurator.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    size_370M = "370M"
+    size_600M = "600M"
+    size_760M = "760M"
+    size_1B = "1B"
+    size_3B = "3B"
+    size_7B = "7B"
+    size_13B = "13B"
+
+    @property
+    def num_params(self) -> int:
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder2/transformer_model_configurator.py]: Slightly confusing that this returns the rough number of params, but there is a similar method `get_num_params` in `base.py` that returns the exact number of non-embedding params.

File: src/olmo_core/model_ladder2/transformer_model_configurator.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        device_type = device_type.lower()
+        assert "h100" in device_type or "b200" in device_type
+        assert sequence_length in {2048, 4096, 8192}
+        size_spec = TransformerSize(size_spec)
+
+        dp_config = TransformerDataParallelConfig(
+            name=DataParallelType.fsdp,
+            param_dtype=DType.bfloat16,
+            reduce_dtype=DType.float32,
+            wrapping_strategy=TransformerDataParallelWrappingStrategy.blocks,
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder2/transformer_model_configurator.py]: nit: I think full wrapping is slightly better than blocks in general. This is because under blocks wrapping the LM head isnt released for communication until after the backward pass is complete.

File: src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    def configure_duration(self, num_params: int) -> Duration:
+        return self._chinchilla_duration(
+            num_params,
+            self.chinchilla_multiple,
+        )
+
+    def configure_target_batch_size(self, num_params: int) -> int:
+        # Calculate global batch size according to https://api.semanticscholar.org/CorpusID:270764838
+        # which assumes a sequence length of 2048.
```
Comment (tyler-romero) [inline src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py:51]: hmmm, any idea if / how this would be modified for 8k seq len?

File: src/scripts/train/ladder/olmo3_ladder.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+            """
+        ).replace("\n", " "),
+        "launch_benchmark": "Launch a Beaker job to run a benchmark for a given model size.",
+        "run": textwrap.dedent(
+            """
+            Run the ladder for a given model size locally.
+            This usually isn't invoked directly, but rather via `launch_run`.
+            """
+        ).replace("\n", " "),
+        "launch_run": "Launch a Beaker job to run the ladder for a given model size.",
```
Comment (tyler-romero) [inline src/scripts/train/ladder/olmo3_ladder.py]: Could we have a utility for launching a set of runs? For example all of the rungs of size less than or equal to the specified size? I see myself kicking off the bottom 3/4 sizes all at once pretty frequently when evaluating ideas.

File: src/scripts/train/ladder/olmo3_ladder.py [hunk 12]
```diff
@@ -0,0 +1,387 @@
+import argparse
```
Comment (tyler-romero) [inline src/scripts/train/ladder/olmo3_ladder.py:1]: So is the expected pattern similar to our other `scripts/train` scripts where users will clone this file and make the changes relevant to their experiment?

If so, could we pull some of the utility functions in this file into a shared file, similar to `internal/experiment.py`?

File: src/olmo_core/model_ladder2/base.py (hunk 6 above)
Comment (epwalsh) [reply src/olmo_core/model_ladder2/base.py]: Done: d65361d8

File: src/olmo_core/model_ladder2/transformer_model_configurator.py (hunk 8 above)
Comment (epwalsh) [reply src/olmo_core/model_ladder2/transformer_model_configurator.py]: Yea... renamed to `approx_num_params` d194910a

File: src/olmo_core/model_ladder2/transformer_model_configurator.py (hunk 9 above)
Comment (epwalsh) [reply src/olmo_core/model_ladder2/transformer_model_configurator.py]: 916779d4

File: src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py (hunk 10 above)
Comment (epwalsh) [reply src/olmo_core/model_ladder/wsds_chinchilla_run_configurator.py:51]: I'm not sure how brittle these scaling equations are to sequence length. We're assuming that 8k is not too different from 2k, so the target batch size in tokens is not a function of sequence length. That seems reasonable enough.

File: src/scripts/train/ladder/olmo3_ladder.py (hunk 11 above)
Comment (epwalsh) [reply src/scripts/train/ladder/olmo3_ladder.py]: 9caf27a6

File: src/olmo_core/model_ladder/base.py [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+        # by a factor that's just the number of gradient accumulation steps needed with the minimum
+        # requested number of devices.
+        max_num_grad_accum_steps = global_batch_size // gbz_factor
+        expansion_factor = min(self.max_devices // min_world_size, max_num_grad_accum_steps)
+        num_devices = min_world_size * expansion_factor
+        dp_world_size = min_dp_world_size * expansion_factor
+
+        # Finally we ensure `global_batch_size` is divisible by the micro-batch size.
+        microbatch_size = rank_microbatch_size * dp_world_size
+        global_batch_size = round(global_batch_size / microbatch_size) * microbatch_size
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply src/olmo_core/model_ladder/base.py]: 12b2bc8f

File: src/olmo_core/internal/ladder.py [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+        max_step_completed = 0
+        max_step = -1
+        checkpoint_displays = []
+        for ckpt in checkpoints:
+            max_step = max(max_step, ckpt.step)
+            if ckpt.exists:
+                max_step_completed = max(max_step_completed, ckpt.step)
+            checkpoint_displays.append(ckpt.display())
+
+        assert max_step > 0
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply src/olmo_core/internal/ladder.py:440]: 12b2bc8f

File: src/olmo_core/internal/ladder.py [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+    size_enum: Type[TransformerSize] = TransformerSize,
+    add_additional_args: Callable[[str, argparse.ArgumentParser], None] | None = None,
+):
+    args = parse_args(
+        configure_ladder, size_enum=size_enum, add_additional_args=add_additional_args
+    )
+    if hasattr(args, "func"):
+        args.func(args)
+    else:
+        args.print_help()
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply src/olmo_core/internal/ladder.py]: 12b2bc8f

File: src/olmo_core/internal/ladder.py [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+
+
+def configure_launcher(
+    args: argparse.Namespace, ladder: ModelLadder, cmd: str, size: str | None = None
+) -> BeakerLaunchConfig:
+    size = size or args.size
+    num_gpus = ladder.get_num_devices(size)
+    assert (num_gpus % 8 == 0) or num_gpus < 8
+    launch_config = build_launch_config(
+        cmd=[sys.argv[0], cmd] + sys.argv[2:],
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply src/olmo_core/internal/ladder.py]: 2a819cd3

File: src/olmo_core/utils.py [hunk 17]
```diff
@@ -723,6 +723,11 @@ def log_once(logger: logging.Logger, msg: str, *args, level: int = logging.INFO,
     logger.log(level, msg, *args, **kwargs)
 
 
+@lru_cache(maxsize=128)
+def warn_once(msg: str, *args, **kwargs):
+    warnings.warn(msg, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply src/olmo_core/utils.py:728]: That's just not true.
